include/triton/Conversion/TritonGPUToLLVM/Utility.h

```diff
@@ -833,16 +854,30 @@ emitOffsetForMfmaLayout(const AMDMfmaEncodingAttr &mfmaLayout,
   auto shapePerCTA = getShapePerCTA(mfmaLayout, tensorShape);
   auto warpsPerCTA = mfmaLayout.getWarpsPerCTA();
 
-  SmallVector<unsigned> numWarpsPerDim(2);
-  for (unsigned d = 0; d < 2; ++d) {
+  auto rank = tensorShape.size();
+
+  SmallVector<unsigned> numWarpsPerDim(rank);
+  unsigned mDim = mfmaLayout.getMDim();
+  unsigned nDim = mfmaLayout.getNDim();
+  assert((mDim == nDim && (mDim == 32 || mDim == 16 || mDim == 4)) ||
+         (mDim == 64 && nDim == 4) || (mDim == 4 && nDim == 64));
+  SmallVector<unsigned> tileShape(rank, 1);
+  tileShape[rank - 2] = mDim;
+  tileShape[rank - 1] = nDim;
+  for (unsigned d = 0; d < rank; ++d) {
     unsigned inPerCTA = std::min<unsigned>(tensorShape[d], shapePerCTA[d]);
     unsigned inPerWarp = ceil<unsigned>(inPerCTA, warpsPerCTA[d]);
-    numWarpsPerDim[d] = ceil<unsigned>(inPerWarp, mfmaLayout.getMDim());
+    numWarpsPerDim[d] = ceil<unsigned>(inPerWarp, tileShape[d]);
   }
 
-  for (unsigned i = 0; i < numWarpsPerDim[0]; ++i) {
-    for (unsigned j = 0; j < numWarpsPerDim[1]; ++j) {
-      emitMfmaOffsetForCTA(mfmaLayout, offsets, i, j);
+  unsigned repBatch = rank == 3 ? numWarpsPerDim[0] : 1;
+  auto warpsPerBatch =
+      rank == 3 ? std::min<unsigned>(tensorShape[0], warpsPerCTA[0]) : 1;
+  for (unsigned b = 0; b < repBatch; ++b) {
+    for (unsigned i = 0; i < numWarpsPerDim[rank - 2]; ++i) {
+      for (unsigned j = 0; j < numWarpsPerDim[rank - 1]; ++j) {
+        emitMfmaOffsetForCTA(mfmaLayout, offsets, b * warpsPerBatch, i, j);
```

Comment:
I'm confused about why do we pass what seems to be the warp index `i` and `j` to the function as `ctaOffsetX` and `ctaOffsetY`.

Reply:
You are right.

Reply:
Is the tileShape here the shape per warp ? 

```
  tileShape[rank - 2] = mDim;
  tileShape[rank - 1] = nDim;
```

Reply:
yes

Reply:
I think so far the terminology is tileShape would be the CTA shape, so probably you might wana call this `shapePerWarp` ?

Reply:
make sense